File: n-doors.py
# The problem is that you have n doors numbered from 1 to n and you have i passes where i = 1,2,3,...,n.
# In the ith pass you open the doors that are multiples of i. What is the number of doors that are open after the last pass?

def n_doors(n):
    doors = n * [0] # Create a list of n doors, all closed
    print('At start        :', doors)
    for i in range(1, n + 1):
        for j in range(i-1, n, i):
            doors[j] = 1 - doors[j] # Toggle the door
        print('When i =', i,'then :', doors)

    open_doors = [k + 1 for k in range(n) if doors[k] == 1] # index + 1 gives the door number

    return doors, sum(doors), open_doors

n = 10
doors, num_open_doors, open_doors = n_doors(n)
print(f"Number of doors: {n}")
print(f"Number of open doors: {num_open_doors}")
print(f"Open doors: {open_doors}")

File: n_doors.js
function nDoors(n) {
  let doors = new Array(n).fill(0); // Create an array of n doors, all closed
  console.log("At start :", doors);
  for (let i = 1; i <= n; i++) {
    for (let j = i - 1; j < n; j += i) {
      doors[j] = 1 - doors[j]; // Toggle the door
    }
    console.log("When i =", i, "then :", doors);
  }

  let openDoors = [];
  for (let k = 0; k < n; k++) {
    if (doors[k] === 1) {
      openDoors.push(k + 1); // Index + 1 gives the door number
    }
  }

  return {
    doors: doors,
    numOpenDoors: doors.reduce((acc, val) => acc + val, 0),
    openDoors: openDoors,
  };
}

let n = 10;
let result = nDoors(n);
console.log("Number of doors: " + n);
console.log("Number of open doors: " + result.numOpenDoors);
console.log("Open doors: " + result.openDoors);
